File: core/runtime.lua/prelude.lua
-- This is free and unencumbered software released into the public domain.

--- DRYlib for Lua
-- @module dry
local dry = {
  geo  = {},
  net  = {},
  std  = {},
  sys  = {},
  time = {},
}

--------------------------------------------------------------------------------
-- Type Constructors

-- Boolean (true or false)
function dry.bool(x)
  return x -- TODO
end

-- Character (21-bit Unicode code point)
function dry.char(c)
  return c -- TODO
end

-- Complex number (arbitrary size)
function dry.complex(real, imaginary)
  return {dry.real(real), dry.real(imaginary)} -- TODO
end

-- Floating-point number (native size)
function dry.float(r)
  return dry.float64(r)
end

-- Floating-point number (32-bit single-precision)
function dry.float32(r)
  return r -- TODO
end

-- Floating-point number (64-bit double-precision)
function dry.float64(r)
  return r -- TODO
end

-- Integer number (native size)
function dry.int(z)
  return dry.int64(z)
end

-- Integer number (8-bit)
function dry.int8(z)
  return z -- TODO
end

-- Integer number (16-bit)
function dry.int16(z)
  return z -- TODO
end

-- Integer number (32-bit)
function dry.int32(z)
  return z -- TODO
end

-- Integer number (64-bit)
function dry.int64(z)
  return z -- TODO
end

-- Integer number (128-bit)
function dry.int128(z)
  return dry.int64(z) -- FIXME
end

-- Integer number (arbitrary size)
function dry.integer(z)
  return dry.int64(z) -- FIXME
end

-- Natural number (arbitrary size)
function dry.natural(n)
  return dry.integer(n) -- TODO
end

-- Rational number (arbitrary size)
function dry.rational(numerator, denominator)
  return {dry.integer(numerator), dry.integer(denominator)} -- TODO
end

-- Real number (arbitrary size)
function dry.real(r)
  return dry.float64(r) -- FIXME
end

-- Machine word (native size)
function dry.word(n)
  return dry.word64(n)
end

-- Machine word (8-bit)
function dry.word8(n)
  return n -- TODO
end

-- Machine word (16-bit)
function dry.word16(n)
  return n -- TODO
end

-- Machine word (32-bit)
function dry.word32(n)
  return n -- TODO
end

-- Machine word (64-bit)
function dry.word64(n)
  return n -- TODO
end

--------------------------------------------------------------------------------

--[[
print("OK") -- lua5.3 prelude.lua
--]]

return dry

File: core/runtime.kt/prelude.kt
/* This is free and unencumbered software released into the public domain. */

package dry

////////////////////////////////////////////////////////////////////////////////
// Type Definitions

// Boolean (true or false)
typealias Bool = kotlin.Boolean

// Character (Unicode code point)
typealias Char = kotlin.Char // FIXME: this is only 16-bit

// Complex number (arbitrary size)
data class Complex(val real: Real, val imaginary: Real)

// Floating-point number (native size)
typealias Float = kotlin.Double

// Floating-point number (32-bit single-precision)
typealias Float32 = kotlin.Float

// Floating-point number (64-bit double-precision)
typealias Float64 = kotlin.Double

// Integer number (native size)
typealias Int = kotlin.Int

// Integer number (8-bit)
typealias Int8 = kotlin.Int

// Integer number (16-bit)
typealias Int16 = kotlin.Int

// Integer number (32-bit)
typealias Int32 = kotlin.Int

// Integer number (64-bit)
typealias Int64 = kotlin.Long

// Integer number (128-bit)
typealias Int128 = kotlin.Long // FIXME

// Integer number (arbitrary size)
typealias Integer = java.math.BigInteger

// Natural number (arbitrary size)
typealias Natural = Integer

// Rational number (arbitrary size)
data class Rational(val numerator: Integer, val denominator: Integer)

// Real number (arbitrary size)
typealias Real = java.math.BigDecimal

// Machine word (native size)
typealias Word = kotlin.Long

// Machine word (8-bit)
typealias Word8 = kotlin.Long

// Machine word (16-bit)
typealias Word16 = kotlin.Long

// Machine word (32-bit)
typealias Word32 = kotlin.Long

// Machine word (64-bit)
typealias Word64 = kotlin.Long

////////////////////////////////////////////////////////////////////////////////

//println("OK") // kotlinc -script prelude.kts
